DataFormats/TrackSoA/interface/TracksSoA.h

```diff
@@ -12,91 +12,141 @@
-  template <typename TrackerTraits>
-  using TrackSoAView = typename reco::TrackSoA<TrackerTraits>::template Layout<>::View;
-  template <typename TrackerTraits>
-  using TrackSoAConstView = typename reco::TrackSoA<TrackerTraits>::template Layout<>::ConstView;
-
-  /* Implement a type trait to identify the specialisations of TrackSoAConstView<TrackerTraits>
-   *
-   * This is done explicitly for all possible pixel topologies, because we did not find a way
-   * to use template deduction with a partial specialisation.
-   */
-  template <typename T>
-  struct IsTrackSoAConstView : std::false_type {};
-  template <>
-  struct IsTrackSoAConstView<TrackSoAConstView<pixelTopology::Phase1>> : std::true_type {};
-  template <>
-  struct IsTrackSoAConstView<TrackSoAView<pixelTopology::Phase1>> : std::true_type {};
-  template <>
-  struct IsTrackSoAConstView<TrackSoAConstView<pixelTopology::Phase2>> : std::true_type {};
-  template <>
-  struct IsTrackSoAConstView<TrackSoAView<pixelTopology::Phase2>> : std::true_type {};
-  template <>
-  struct IsTrackSoAConstView<TrackSoAConstView<pixelTopology::HIonPhase1>> : std::true_type {};
-  template <>
-  struct IsTrackSoAConstView<TrackSoAView<pixelTopology::HIonPhase1>> : std::true_type {};
-
-  template <typename T>
-  constexpr bool isTrackSoAConstView = IsTrackSoAConstView<T>::value;
-
-  template <typename ConstView, typename = std::enable_if_t<isTrackSoAConstView<ConstView>>>
-  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr float charge(ConstView const& tracks, int32_t i) {
+  // Aliases in order to not confuse the GENERATE_SOA_LAYOUT
+  // macro with weird colons and angled brackets.
+  using Vector5f = Eigen::Matrix<float, 5, 1>;
+  using Vector15f = Eigen::Matrix<float, 15, 1>;
+  using Quality = pixelTrack::Quality;
+
+  GENERATE_SOA_LAYOUT(TrackLayout,
+                      SOA_COLUMN(Quality, quality),
+                      SOA_COLUMN(float, chi2),
+                      SOA_COLUMN(int8_t, nLayers),
+                      SOA_COLUMN(float, eta),
+                      SOA_COLUMN(float, pt),
+                      // state at the beam spot: {phi, tip, 1/pt, cotan(theta), zip}
+                      SOA_EIGEN_COLUMN(Vector5f, state),
+                      SOA_EIGEN_COLUMN(Vector15f, covariance),
+                      SOA_SCALAR(int, nTracks),
+                      SOA_COLUMN(uint32_t, hitOffsets))
+
+  GENERATE_SOA_LAYOUT(TrackHitsLayout, SOA_COLUMN(uint32_t, id), SOA_COLUMN(uint32_t, detId))
+
+  using TrackSoA = TrackLayout<>;
+  using TrackSoAView = TrackSoA::View;
+  using TrackSoAConstView = TrackSoA::ConstView;
+
+  using TrackHitSoA = TrackHitsLayout<>;
+  using TrackHitSoAView = TrackHitSoA::View;
+  using TrackHitSoAConstView = TrackHitSoA::ConstView;
+
+  // All these below were constexpr. Now I get this:
+  // note: non-literal type 'reco::TrackLayout<128, false>::ConstViewTemplateFreeParams<128, false, true, true>::const_element'
```

Comment:
Could it be because before the types were still templated, while now they are fully specialised ?

Reply:
@leobeltra could you remind me what `const_element` expands to for this SoA ?

Reply:
Thanks.

I guess it makes sense that `element` and `const_element` are not ["literal types"](https://en.cppreference.com/w/cpp/named_req/LiteralType).

Maybe it would be enough to mark the constructor of `element` and `const_element` as `constexpr` ?

Reply:
@AdrianoDee could you check if https://github.com/cms-sw/cmssw/pull/48103 lets you mark these functions as `constexpr` ?

Reply:
OK, so we fixed our side of things, and now are hitting non-`constexpr` code in Eigen.